Feat #5968/forgot password (#5971)
vuestorefront/vue-storefront: packages/core/docs/commercetools/composables/use-forgot-password.md

# `useForgotPassword`

## Features

`useForgotPassword` composable can be used to:

* generate reset password token
* set new password using token

## API

> - `reset` - function used to generate reset password token. When invoked, it requests data from the API and populates `token` property. This method accepts a single params object. The `params` has the following option:

Reviewer: ```suggestion
- `reset` - function to generate reset password token. When invoked, it requests data from the API and populates the `token` property. This method accepts a single `params` object with the following properties:
```

  - `params: ResetPasswordParams`

  - `customQuery?: CustomQuery`

```typescript
interface ResetPasswordParams {
    email: string;
}

type CustomQuery = {
  customerCreatePasswordResetToken: string
}
```

> - `setNew` - function used to set new user password after reset. When invoked, it requests data from the API and populates `result` property. This method accepts a single params object. The `params` has the following option:

Reviewer: ```suggestion
- `setNew` - function to set new user password after `reset`. When invoked, it requests data from the API and populates the `result` property. This method accepts a single `params` object with the following properties:
```

  - `params: SetNewPasswordParams`
  
  - `customQuery?: CustomQuery`

```typescript
interface SetNewPasswordParams {
    tokenValue: string;
    newPassword: string;
}

type CustomQuery = {
  customerResetPassword: string
}
```

> - `token: string` - reactive data string containing the created token.

Reviewer: ```suggestion
- `token: string` - reactive data string containing the reset token.
```

- `loading: boolean` - reactive object containing information about loading state of `setNew` and `reset` methods.

- `error: UseForgotPasswordErrors` - reactive object containing the error message, if `setNew` or `reset` failed for any reason.

```ts
interface UseForgotPasswordErrors {
  reset: Error;
  setNew: Error;
}
```

## Getters

> We do not provide getters for `useForgotPassword` composable.

Reviewer: ```suggestion
There are no getters for `useForgotPassword` composable.
```

```ts
interface UseForgotPasswordErrors {
  request: Error;
  setNew: Error;
}
```

## Getters

- `getResetPasswordToken` - returns generated reset password token.

- `isPasswordChanged` - returns a boolean value of a password set status.

```ts
interface ForgotPasswordGetters<ForgotPasswordResult> {
  getResetPasswordToken: (result: ForgotPasswordResult) => string
  isPasswordChanged: (result: ForgotPasswordResult) => boolean
}

interface ForgotPasswordResult {
  resetPasswordResult: CreatePasswordResetTokenResponse;
  setNewPasswordResult: ResetPasswordResponse;
}

type CreatePasswordResetTokenResponse = QueryResponse<'customerCreatePasswordResetToken', CustomerPasswordToken>;

type ResetPasswordResponse = QueryResponse<'customerResetPassword', Customer>;

type CustomerPasswordToken = {
  customerId: Scalars["String"];
  expiresAt: Scalars["DateTime"];
  value: Scalars["String"];
  id: Scalars["String"];
  version: Scalars["Long"];
  createdAt: Scalars["DateTime"];
  lastModifiedAt: Scalars["DateTime"];
  createdBy?: Maybe<Initiator>;
  lastModifiedBy?: Maybe<Initiator>;

> }

Reviewer: This should come from API reference. Please ask @andrzejewsky to generate new GraphQL types and generate new API reference. Then link to instead of pasting whole interface

Author: Created Jira issue for that as this recommendation is out of the forgot password and will be needed in other places as well.

Reviewer 2: @filipsobol completely right, but let's do this in a different task, because it sound like other scope

  firstName?: Maybe<Scalars["String"]>;
  lastName?: Maybe<Scalars["String"]>;
  middleName?: Maybe<Scalars["String"]>;
  title?: Maybe<Scalars["String"]>;
  locale?: Maybe<Scalars["Locale"]>;
  salutation?: Maybe<Scalars["String"]>;
  dateOfBirth?: Maybe<Scalars["Date"]>;
  companyName?: Maybe<Scalars["String"]>;
  vatId?: Maybe<Scalars["String"]>;
  customerGroup?: Maybe<CustomerGroup>;
  defaultShippingAddress?: Maybe<Address>;
  defaultBillingAddress?: Maybe<Address>;
  shippingAddresses: Array<Address>;
  billingAddresses: Array<Address>;
  storesRef: Array<KeyReference>;
  stores: Array<Store>;
  /** This field contains non-typed data. Consider using `customFields` as a typed alternative. */
  customFieldsRaw?: Maybe<Array<RawCustomField>>;
  /** This field would contain type data */
  customFields?: Maybe<Type>;
  custom?: Maybe<CustomFieldsType>;
  id: Scalars["String"];
  version: Scalars["Long"];
  createdAt: Scalars["DateTime"];
  lastModifiedAt: Scalars["DateTime"];
  createdBy?: Maybe<Initiator>;
  lastModifiedBy?: Maybe<Initiator>;
  /** Custom fields are returned as a list instead of an object structure. */
  customFieldList?: Maybe<Array<CustomField>>;

> };

Reviewer: Link to `commercetools/api-client-reference.html#customer` instead of duplicating the whole interface

  - `params: IResetPasswordParams`

  - `customQuery?: CustomQuery`

```typescript
interface IResetPasswordParams {
    email: string;
}

type CustomQuery = {
  customerCreatePasswordResetToken: string
}
```

- `change` - function used to set new user password after reset. When invoked, it requests data from the API and populates `result` property. This method accepts a single params object. The `params` has the following option:

  - `params: IChangePasswordParams`
  
  - `customQuery?: CustomQuery`

```typescript
interface IChangePasswordParams {
    tokenValue: string;
    newPassword: string;
}

type CustomQuery = {
  customerResetPassword: string
}
```

- `token: string` - reactive data string containing the created token.

- `loading: boolean` - reactive object containing information about loading state of `change` and `reset` methods.

- `error: UseForgotPasswordErrors` - reactive object containing the error message, if `change` or `reset` failed for any reason.

```ts
interface UseForgotPasswordErrors {

>   result: Error;

Author: As the two methods cannot be called in one step (you genereate reset token in one page and change password in another) there was no need for me to generate separate keys for errors in these methods. This results in simpler template (as we do not need to create separate conditionals for errors, we just have one) and in less logic because there is only one error. Do you think there should be separate errors for these methods?

}
```

## Getters

We do not provide getters for forgotPassword and its parts.

## Example

Generating reset password token and changing user password.

```vue
<template>
  <!-- Generate reset password token -->
  <form @submit.prevent="reset({ email: form.value.email })">
    <!-- form fields -->
    <button type="submit" :disabled="loading">Reset Password</button>
  </form>

  <!-- Change user password -->
  <form @submit.prevent="change({ tokenValue: <TOKEN_FROM_URL>, newPassword: form.value.password })">
    <!-- form fields -->
    <button type="submit" :disabled="loading">Save Password</button>
  </form>
  <div>{{ result }} - Boolean confirmation of successful password change</div>
</template>

<script>
import { useForgotPassword } from '@vsf-enterprise/commercetoolss';

export default {
  setup() {
    const { reset, change, result, loading } = useForgotPassword();

    return {
      reset,
      change,
      result,
      loading,
    };
  }
};

> </script>

Reviewer: I have some concerns about this code example:

1. I feel like we are mixing two things in one code block. Does it make sense to split this example into two parts - one example for generating token and the other for changing user password?
2. I'm not sure what `<TOKEN_FROM_URL>` is and where it comes from. Does calling `reset` triggers an email that contains a link with the token?
3. You are referring to the `form` object that isn't registered and resulted from the `setup` method.

Author: 1. We have an example in authentication section in general docs about forgot password and there are two separate code blocks. That is why I wanted to do something different here that just adding path to `@vsf-enterprise/commercetools`.
2. Reset method does not trigger an email because we do not have an email provider integrated yet. We just call a GraphQL mutation that returns a token and we log it to the console for now to simulate real life example. I can change the TOKEN_FROM_URL to how it is implemented.
3. Yes, I wanted to not include too many things at once. But now when I look at it, it may look strange (Where does it come from?)
